pallets/pool-system/src/tranches.rs

```diff
@@ -1394,12 +1577,2336 @@ where
+
+			let rev_ids: Vec<TrancheId> = tranches.ids.clone().into_iter().rev().collect();
+			assert_eq!(rev_ids, tranches.ids_non_residual_top());
+		}
+
+		#[test]
+		fn combine_residual_top_tranches_works() {
+			let mut i: Seniority = 0;
+			assert_eq!(
+				default_tranches()
+					.combine_with_residual_top(&[220, 210, 250], |tranche, other_val| {
+						i += 1;
+						Ok((tranche.seniority + i, *other_val))
+					})
+					.unwrap(),
+				[(1, 220), (2, 210), (3, 250)]
+			);
+
+			// tranches has smaller size than combinator
+			assert_eq!(
+				default_tranches()
+					.combine_with_residual_top(&[220, 210, 250, 110], |tranche, other_val| {
+						Ok((tranche.seniority, *other_val))
+					}),
+				Err(DispatchError::Other(
+					"Iterable contains more elements than Tranches tranche count",
+				))
+			);
+
+			// tranches has greater size than combinator
+			assert_eq!(
+				default_tranches().combine_with_residual_top(&[220, 210], |tranche, other_val| {
+					Ok((tranche.seniority, *other_val))
+				}),
+				Err(DispatchError::Other(
+					"Tranches struct contains more tranches than combining iterable"
+				))
+			);
+
+			// combinator is empty
+			assert_eq!(
+				default_tranches()
+					.combine_with_residual_top(vec![], |tranche, _: u32| { Ok(tranche.seniority) }),
+				Err(DispatchError::Other(
+					"Tranches struct contains more tranches than combining iterable"
+				))
+			);
+		}
+
+		#[test]
+		fn combine_with_mut_residual_top_tranches_works() {
+			let mut tranches = default_tranches();
+			let values = [10, 20, 30];
+			assert_eq!(
+				tranches.combine_with_mut_residual_top(&values, |tranche, new_seniority| {
+					tranche.seniority = *new_seniority;
+					Ok(tranche.seniority)
+				}),
+				Ok(values.into())
+			);
+
+			// tranches has smaller size than combinator
+			let values_too_many = [10, 20, 30, 40];
+			assert_eq!(
+				tranches.combine_with_mut_residual_top(&values_too_many, |tranche, other_val| {
+					Ok(tranche.seniority + other_val)
+				}),
+				Err(DispatchError::Other(
+					"Iterable contains more elements than Tranches tranche count",
+				))
+			);
+
+			// tranches has greater size than combinator
+			let mut tranches = default_tranches();
+			let values_too_few = [10, 20];
+			assert_eq!(
+				tranches.combine_with_mut_residual_top(&values_too_few, |tranche, other_val| {
+					Ok(tranche.seniority + other_val)
+				}),
+				Err(DispatchError::Other(
+					"Tranches struct contains more tranches than combining iterable"
+				))
+			);
+
+			// combinator is empty
+			assert_eq!(
+				tranches.combine_with_mut_residual_top(vec![], |tranche, _: u32| {
+					Ok(tranche.seniority)
+				}),
+				Err(DispatchError::Other(
+					"Tranches struct contains more tranches than combining iterable"
+				))
+			);
+		}
+
+		mod calculate_prices {
+			use super::*;
+
+			// FIXME: To reviewer, can this somehow be derived? In practice, we only need the `total_issuance` but `calculate_price` requires the trait.
```

Comment:
I don't think so. We will in the near future add an issuance field to the tranches. As we need to ensure that ONLY the pool is able to increase/decrease those values, to protect us against minting bugs at the bridge levels.